File: backend/app/models/question.py
from typing import Optional

from sqlalchemy import Boolean, ForeignKey, Index, Integer, String, Text
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.db.base import Base, TimestampMixin

DIFFICULTIES = ("easy", "medium", "hard")


class Question(Base, TimestampMixin):
    __tablename__ = "questions"
    __table_args__ = (Index("ix_questions_filters", "subject_id", "chapter_id", "topic_id", "difficulty"),)

    id: Mapped[int] = mapped_column(primary_key=True)
    subject_id: Mapped[int] = mapped_column(ForeignKey("subjects.id", ondelete="CASCADE"), index=True)
    chapter_id: Mapped[int] = mapped_column(ForeignKey("chapters.id", ondelete="CASCADE"), index=True)
    topic_id: Mapped[int | None] = mapped_column(
        ForeignKey("topics.id", ondelete="SET NULL"), nullable=True, index=True
    )

    text: Mapped[str] = mapped_column(Text, nullable=False)
    explanation: Mapped[str] = mapped_column(Text, default="")
    difficulty: Mapped[str] = mapped_column(String(10), default="medium", index=True)
    source: Mapped[str] = mapped_column(String(120), default="")
    year: Mapped[int | None] = mapped_column(Integer, nullable=True)
    image_url: Mapped[str | None] = mapped_column(String(500), nullable=True)
    is_active: Mapped[bool] = mapped_column(Boolean, default=True, nullable=False)
    created_by: Mapped[int | None] = mapped_column(ForeignKey("users.id", ondelete="SET NULL"), nullable=True)

    options: Mapped[list["QuestionOption"]] = relationship(
        back_populates="question", cascade="all, delete-orphan", order_by="QuestionOption.label"
    )

    @property
    def correct_option(self) -> Optional["QuestionOption"]:
        return next((option for option in self.options if option.is_correct), None)


class QuestionOption(Base):
    __tablename__ = "question_options"
    __table_args__ = (Index("ix_question_options_question", "question_id", "label"),)

    id: Mapped[int] = mapped_column(primary_key=True)
    question_id: Mapped[int] = mapped_column(ForeignKey("questions.id", ondelete="CASCADE"), index=True)
    label: Mapped[str] = mapped_column(String(2), nullable=False)
    text: Mapped[str] = mapped_column(Text, nullable=False)
    is_correct: Mapped[bool] = mapped_column(Boolean, default=False, nullable=False)

    question: Mapped[Question] = relationship(back_populates="options")

File: mobile/lib/models/question.dart
class QuestionOption {
  QuestionOption({required this.id, required this.label, required this.text});

  factory QuestionOption.fromJson(Map<String, dynamic> json) => QuestionOption(
        id: json['id'] as int,
        label: json['label'] as String,
        text: json['text'] as String,
      );

  final int id;
  final String label;
  final String text;
}

class Question {
  Question({
    required this.id,
    required this.text,
    required this.difficulty,
    required this.subjectId,
    required this.chapterId,
    required this.options,
    required this.isBookmarked,
    this.topicId,
    this.source = '',
    this.year,
    this.imageUrl,
  });

  factory Question.fromJson(Map<String, dynamic> json) => Question(
        id: json['id'] as int,
        text: json['text'] as String,
        difficulty: json['difficulty'] as String? ?? 'medium',
        subjectId: json['subject_id'] as int,
        chapterId: json['chapter_id'] as int,
        topicId: json['topic_id'] as int?,
        source: json['source'] as String? ?? '',
        year: json['year'] as int?,
        imageUrl: json['image_url'] as String?,
        options: ((json['options'] as List<dynamic>?) ?? [])
            .map((item) => QuestionOption.fromJson(item as Map<String, dynamic>))
            .toList(),
        isBookmarked: json['is_bookmarked'] as bool? ?? false,
      );

  final int id;
  final String text;
  final String difficulty;
  final int subjectId;
  final int chapterId;
  final int? topicId;
  final String source;
  final int? year;
  final String? imageUrl;
  final List<QuestionOption> options;
  final bool isBookmarked;
}
